src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs

```diff
@@ -137,17 +131,138 @@ internal class GrpcWorkerChannel : IRpcWorkerChannel, IDisposable
 
         internal RpcWorkerConfig Config => _workerConfig;
 
+        private void ProcessItem(InboundGrpcEvent msg)
+        {
+            // note this is a QUWI entry-point
+            try
+            {
+                switch (msg.MessageType)
+                {
+                    case MsgType.RpcLog:
+                        switch (msg.Message.RpcLog.LogCategory)
+                        {
+                            case RpcLogCategory.System:
+                                SystemLog(msg);
+                                break;
+                            default:
+                                Log(msg);
+                                break;
+                        }
+                        break;
+                    case MsgType.WorkerStatusResponse:
+                        ReceiveWorkerStatusResponse(msg.Message.RequestId, msg.Message.WorkerStatusResponse);
+                        break;
+                    case MsgType.InvocationResponse:
+                        _ = InvokeResponse(msg.Message.InvocationResponse);
+                        break;
+                    default:
+                        OnNext(msg);
+                        break;
+                }
+            }
+            catch
+            {
+                // TODO: log?
+            }
+        }
+
+        private void OnNext(InboundGrpcEvent message)
+        {
+            Queue<PendingItem> queue;
+            lock (_pendingActions)
+            {
+                if (!_pendingActions.TryGetValue(message.MessageType, out queue))
+                {
+                    return; // nothing to do
+                }
+            }
+            PendingItem next;
+            lock (queue)
+            {
+                do
+                {
+                    if (!queue.TryDequeue(out next))
+                    {
+                        return; // nothing to do
+                    }
+                }
+                while (next.IsComplete);
+            }
+            next.SetResult(message);
+        }
+
+        private void OnNext(MsgType messageType, TimeSpan timeout, int count, Action<InboundGrpcEvent> callback, Action<Exception> faultHandler)
+        {
+            Queue<PendingItem> queue;
+            lock (_pendingActions)
+            {
+                if (!_pendingActions.TryGetValue(messageType, out queue))
+                {
+                    queue = new Queue<PendingItem>();
+                    _pendingActions.Add(messageType, queue);
+                }
+            }
+
+            lock (queue)
+            {
+                // while we have the lock, discard any dead items (to prevent unbounded growth on stall)
+                while (queue.TryPeek(out var next) && next.IsComplete)
+                {
+                    queue.Dequeue();
+                }
+                for (int i = 0; i < count; i++)
+                {
+                    var newItem = i == count - 1 && timeout != TimeSpan.Zero
```

Comment:
nit: for readability
```suggestion
                    var newItem = (i == count - 1) && (timeout != TimeSpan.Zero)
```